core/src/test/java/edu/wpi/grip/core/sources/HttpSourceTest.java

```diff
@@ -0,0 +1,98 @@
+
+package edu.wpi.grip.core.sources;
+
+import com.google.common.eventbus.EventBus;
+
+import edu.wpi.grip.core.MockPipeline;
+import edu.wpi.grip.core.sockets.OutputSocket;
+import edu.wpi.grip.core.http.GripServer;
+import edu.wpi.grip.core.http.GripServerTest;
+import edu.wpi.grip.core.util.MockExceptionWitness;
+import edu.wpi.grip.util.Files;
+
+import org.apache.commons.httpclient.URIException;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.entity.FileEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.bytedeco.javacpp.opencv_core.Mat;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URISyntaxException;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
+ *
+ */
+public class HttpSourceTest {
+
+    private File logoFile;
+
+    private GripServer server;
+    private HttpSource source;
+    private CloseableHttpClient postClient;
+
+    @Before
+    public void setup() throws URIException, URISyntaxException {
+        GripServer.HttpServerFactory f = new GripServerTest.TestServerFactory();
+        MockPipeline.MockProjectSettings projectSettings = new MockPipeline.MockProjectSettings();
+        projectSettings.setServerPort(8080);
+        MockPipeline pipeline = new MockPipeline(projectSettings);
+        server = GripServerTest.makeServer(f, pipeline);
+        server.start();
+        EventBus eventBus = new EventBus();
+        source = new HttpSource(origin -> new MockExceptionWitness(eventBus, origin), eventBus, server);
+
+        logoFile = new File(Files.class.getResource("/edu/wpi/grip/images/GRIP_Logo.png").toURI());
+        postClient = HttpClients.createDefault();
+    }
+
+    @Test
+    public void testPostImage() throws IOException, InterruptedException {
+        server.addPostHandler(GripServer.IMAGE_UPLOAD_PATH, source);
+        OutputSocket<Mat> imageSource = source.getOutputSockets()[0];
+
+        // We have to manually update the output sockets to get the image
+        source.updateOutputSockets();
+        assertTrue("The value should not be present if the source hasn't been initialized and no image POSTed", imageSource.getValue().get().empty());
+
+        source.initialize(); // adds the source as a PostHandler to the server
+        source.updateOutputSockets();
+        assertTrue("The value should not be present since the source has been initialized but no image POSTed", imageSource.getValue().get().empty());
+
+        doPost(GripServer.IMAGE_UPLOAD_PATH, logoFile);
+        Thread.sleep(100); // wait for the image to be uploaded
```

Comment:
Can you just add a responder that waits for the callback to happen from the HTTP server when the connection happens. Adding waits can slow down tests.
We have a deferred testing library that allows for waiting for callbacks to execute.
This is the library.